File: Python-3.7/data_structures/queue/queue_linked_list.py
"""
OPERATIONS ON QUEUE

- ENQUEUE: Add an item in the queue. If queue is full then it is said to be Stack Overflow.
- DEQUEUE: Remove an item from the queue. If the queue is empty, then it is said to be an Underflow condition.
- FRONT: Return front element.
- REAR: Return rear element.
- ISFULL: Check if queue is full.
- ISEMPTY: Check if queue is empty.

TIME COMPLEXITY: 

- ENQUEUE: O(1)
- DEQUEUE: O(1)
- FRONT: O(1)
- REAR: O(1)
- ISFULL: O(1)
- ISEMPTY: O(1)

APPLICATIONS:

- Breadth First Search.
- CPU Scheduling.
- Disk Scheduling.
- IO Buffers.
- Pipes.
- File IO

IMPLEMENTATION WAYS:

- Using array
- Using linked-list
"""

# import modules
from sys import maxsize

"""
Implementation using Linked List

PROS:
- Easy to implement
- Memory saved as no pointers are involved

CONS:
- Not dynamic
"""
# Queue Node
class Node():

    # COnstructor
    def __init__(self, data):
        self.info = data
        self.next = None

# Queue
class Queue():

    # constructor
    # initialize front, rear, size, queue and capacity
    def __init__(self):
        self.front = None
        self.rear = None
    
    # Function to check if queue is empty
    # Returns: Bool
    def is_empty(self):
        return self.front == None
    
    # Function to add item in queue
    # Returns: item inserted in queue
    def en_queue(self, item):
        
        new_node = Node(item)

        if self.is_empty():
            self.front = self.rear = new_node
            return new_node.info
        
        self.rear.next = new_node
        self.rear = new_node
        return self.rear.info
    
    # Function to remove item in queue
    # Returns: (-inf) if underflow else item removed from queue
    def de_queue(self):
        if self.is_empty():
            # return minus infinite
            return str(-maxsize-1)

        temp = self.front
        self.front = self.front.next

        if self.front == None:
            self.rear = None
        return temp.info
    
    # Function to get front element from queue
    # Returns: (-inf) if queue is empty else front item in queue
    def get_front(self):
        if self.is_empty():
            # return minus infinite
            return str(-maxsize-1)
        return self.front.info
    
    # Function to get rear element from queue
    # Returns: (-inf) if queue is empty else rear item in queue
    def get_rear(self):
        if self.is_empty():
            # return minus infinite
            return str(-maxsize-1)
        return self.rear.info


# Driver Code to test above class 
# if __name__ == '__main__': 
  
#     # create Queue class object
#     queue = Queue() 

#     print('Queue: Empty ? ', queue.is_empty())
    
#     # add 10 to queue
#     print('inserting 10')
#     queue.en_queue(10) 
#     print('Front: ', queue.get_front())
#     print('Rear: ', queue.get_rear())
    
#     # add 20 to queue
#     print('inserting 20')
#     queue.en_queue(20) 
#     print('Front: ', queue.get_front())
#     print('Rear: ', queue.get_rear())
    
#     # add 30 to queue
#     print('inserting 30')
#     queue.en_queue(30) 
#     print('Front: ', queue.get_front())
#     print('Rear: ', queue.get_rear())
    
#     # add 40 to queue
#     print('inserting 40')
#     queue.en_queue(40) 
#     print('Front: ', queue.get_front())
#     print('Rear: ', queue.get_rear())
    
#     # remove item from queue
#     print('removing first item')
#     queue.de_queue()
#     print('Front: ', queue.get_front())
#     print('Rear: ', queue.get_rear())

File: C++/data_structures/queue/queue_linked_list.cpp
/*
OPERATIONS ON QUEUE

- ENQUEUE: Add an item in the queue. If queue is full then it is said to be Stack Overflow.
- DEQUEUE: Remove an item from the queue. If the queue is empty, then it is said to be an Underflow condition.
- FRONT: Return front element.
- REAR: Return rear element.
- ISFULL: Check if queue is full.
- ISEMPTY: Check if queue is empty.

TIME COMPLEXITY: 

- ENQUEUE: O(1)
- DEQUEUE: O(1)
- FRONT: O(1)
- REAR: O(1)
- ISFULL: O(1)
- ISEMPTY: O(1)

APPLICATIONS:

- Breadth First Search.
- CPU Scheduling.
- Disk Scheduling.
- IO Buffers.
- Pipes.
- File IO

IMPLEMENTATION WAYS:

- Using array
- Using linked-list
*/

#include<bits/stdc++.h>
using namespace std;

class Node {
    public:
        int info;
        Node * next;

        Node(int item) {
            info = item;
            next = NULL;
        }
};

class Queue {

    Node *front, *rear;
    unsigned int capacity, size;
    
    public:

        Queue(unsigned int limit) {
            
            front = NULL;
            rear = NULL;
            capacity = limit;
            size = 0;
        }

        bool is_empty();
        bool is_full();
        int en_queue(int item);
        int de_queue();
        int get_front();
        int get_rear();

};

bool Queue::is_empty() {
    return front == NULL;
}

bool Queue::is_full() {
    return capacity == size;
}

int Queue::en_queue(int item) {

    if(is_full()) {
        return INT_MIN;
    }

    Node * new_node = new Node(item);

    if(is_empty()) {
        front = rear = new_node;
        size += 1;
        return rear->info;
    }
    
    rear->next = new_node;
    rear = new_node;
    size += 1;
    return rear->info;

}

int Queue::de_queue() {

    if(is_empty()) {
        return INT_MIN;
    }

    Node * temp = front;
    front = front->next;

    if(front == NULL) {
        rear = NULL;
    }
    
    size--;
    return temp->info;

}

int Queue::get_front() {
    if(is_empty()) {
        return INT_MIN;
    }
    return front->info;
}

int Queue::get_rear() {
    if(is_empty()) {
        return INT_MIN;
    }
    return rear->info;
}

// Driver code to test above class
// int main()  
// {  
//     Queue queue = Queue(100000);  
  
//     cout << "EN-QUEUE: " << queue.en_queue(10) << endl;  
//     cout << "Front item is " << queue.get_front() << endl;  
//     cout << "Rear item is " << queue.get_rear() << endl;  
    
//     cout << "EN-QUEUE: " << queue.en_queue(20) << endl;  
//     cout << "Front item is " << queue.get_front() << endl;  
//     cout << "Rear item is " << queue.get_rear() << endl;  
    
//     cout << "EN-QUEUE: " << queue.en_queue(30) << endl;  
//     cout << "Front item is " << queue.get_front() << endl;  
//     cout << "Rear item is " << queue.get_rear() << endl;  
  
//     cout << "EN-QUEUE: " << queue.en_queue(40) << endl;  
//     cout << "Front item is " << queue.get_front() << endl;  
//     cout << "Rear item is " << queue.get_rear() << endl;  
  
//     cout << queue.de_queue() << " dequeued from queue\n";  
//     cout << "Front item is " << queue.get_front() << endl;  
//     cout << "Rear item is " << queue.get_rear() << endl;  
  
//     return 0;  
// }  